### PR TITLE
Move psutil to test requirements and do optimistic load

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/10342?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10342/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10342
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
`psutil` doesn't provide Arm64 Linux binary, requiring building from CPython source which is a headache. This change makes it optional and show user-friendly warning when user tries to run system metrics related features without installing it.

Error message will look like this;
```
  File "mlflow/utils/process.py", line 224, in <module>
    _check_and_import_psutil()
  File "mlflow/utils/process.py", line 172, in _check_and_import_psutil
    raise ImportError(
ImportError: No module named 'psutil'

HINT: The `psutil` package is required to log system metrics in MLflow. Run 
`pip install psutil --upgrade` to install the package. If you are on an Arm64-based
 Linux system, you may need to have GCC installed to build the wheel from source.
```

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [x] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -104,17 +106,16 @@ def pytest_runtest_setup(item):
 def pytest_report_teststatus(report, config):
     outcome = yield
     if report.when == "call":
-        try:
-            import psutil
-        except ImportError:
+        # SKip showing stats at all if psutil is not installed
+        if importlib.util.find_spec("psutil") is None:
```
Comment (B-Step62) [inline conftest.py:110]: suggested by ruff

**Comment by harupy** [comment]
We should also think about how we can prevent using `psutil` directly. For example, if I make the following change, would CI block my PR or not? 

```diff
+ import psutil
+
+ print(psutil.__version__)
```

**Comment by B-Step62** [comment]
> We should also think about how we can prevent using `psutil` directly. For example, if I make the following change, would CI block my PR or not?
> 
> ```diff
> + import psutil
> +
> + print(psutil.__version__)
> ```

Makes sense, probably adding custom pylint checker would be the easiest path. Let me give it a try.

**Comment by harupy** [comment]
Have we tried using `LazyLoader`?

```python
from mlflow.utils.lazy_load import LazyLoader

psutil = LazyLoader("psutil", globals(), "psutil")

print(psutil.cpu_count())
```

**Comment by B-Step62** [comment]
> Have we tried using `LazyLoader`?

Ah... didn't know that class. This solves the issue then.

**Comment by B-Step62** [comment]
> Have we tried using `LazyLoader`?
> 
> ```python
> from mlflow.utils.lazy_load import LazyLoader
> 
> psutil = LazyLoader("psutil", globals(), "psutil")
> 
> print(psutil.cpu_count())
> ```

Ah.... actually `psutil` usage is pretty much inside `system_metric` module and `gateway` outside tests.... so it sounds solve the issue.

**Comment by B-Step62** [comment]
> > Have we tried using `LazyLoader`?
> > ```python
> > from mlflow.utils.lazy_load import LazyLoader
> > 
> > psutil = LazyLoader("psutil", globals(), "psutil")
> > 
> > print(psutil.cpu_count())
> > ```
> 
> Ah.... actually `psutil` usage is pretty much inside `system_metric` module and `gateway` outside tests.... so it sounds solve the issue at all.

Just found that `SystemMonitorMetrics` is lazy loaded in current implementation anyway ([code](https://github.com/mlflow/mlflow/blob/e90a2c8ee0160006a83930f68899d8771e381877/mlflow/tracking/fluent.py#L384C1-L385C1)), so probably we don't need to use LazyLoader for it but just keep it and add some helpful message there about psutil.

File: requirements/test-requirements.txt [hunk 2]
```diff
@@ -42,3 +42,5 @@ databricks-sdk
 # Required for LLM eval in `mlflow.evaluate`
 # TODO: Unpin once we support openai 1.0
 openai<1.0
+# Required for showing pytest stats
+psutil<6
```
Comment (harupy) [inline requirements/test-requirements.txt:46]:
```suggestion
psutil
```

File: mlflow/tracking/fluent.py [hunk 3]
```diff
@@ -380,6 +380,17 @@ def start_run(
             # if `log_system_metrics` is not specified, we will check environment variable.
             log_system_metrics = MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING.get()
         if log_system_metrics:
+            # Ensure psutil is installed. It was moved to an optional dependency, as it doesn't
+            # have binary for Arm64 Linux and requires build from CPython which was a headache.
```
Comment (harupy) [inline mlflow/tracking/fluent.py:384]: Any issue we can link?

File: mlflow/tracking/fluent.py [hunk 4]
```diff
@@ -380,6 +380,17 @@ def start_run(
...
+            if importlib.util.find_spec("psutil") is None:
+                raise MlflowException(
+                    "Failed to start system metrics monitoring as package `psutil` is not "
+                    "installed. Run `pip install psutil --upgrade` to resolve the issue, "
```
Comment (harupy) [inline mlflow/tracking/fluent.py:388]:
```suggestion
                    "installed. Run `pip install psutil` to resolve the issue, "
```

File: requirements/test-requirements.txt (hunk 2 above)
Comment (B-Step62) [reply requirements/test-requirements.txt:46]: Got it, will remove

File: mlflow/tracking/fluent.py [hunk 5]
```diff
@@ -380,6 +380,17 @@ def start_run(
...
+                    "otherwise you can disable system metrics logging by passing "
+                    "`log_system_metrics=False` to start_run() or setting environment "
+                    "variable MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING to False."
```
Comment (harupy) [inline mlflow/tracking/fluent.py:391]:
```suggestion
                    "environment variable MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING to False."
```

Comment (B-Step62) [reply mlflow/tracking/fluent.py:391]: oh do you mean changing line break?

File: mlflow/tracking/fluent.py (hunk 4 above)
Comment (B-Step62) [reply mlflow/tracking/fluent.py:388]: ah we don't need, just a random sugestion from copilot

File: mlflow/tracking/fluent.py (hunk 5 above)
Comment (harupy) [reply mlflow/tracking/fluent.py:391]: Never mind. We already have `environment`. I missed it.

**Review by harupy — APPROVED** [review]
LGTM

File: mlflow/tracking/fluent.py [hunk 6]
```diff
@@ -380,6 +380,18 @@ def start_run(
...
+            # Ensure psutil is installed. It was moved to an optional dependency, as it doesn't
+            # have binary for Arm64 Linux and requires build from CPython which was a headache.
+            # https://github.com/giampaolo/psutil/issues/1972
+            if importlib.util.find_spec("psutil") is None:
+                raise MlflowException(
+                    "Failed to start system metrics monitoring as package `psutil` is not "
+                    "installed. Run `pip install psutil` to resolve the issue, "
+                    "otherwise you can disable system metrics logging by passing "
+                    "`log_system_metrics=False` to start_run() or setting environment "
+                    "variable MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING to False."
```
Comment (harupy) [inline mlflow/tracking/fluent.py:392]:
```suggestion
                    f"variable {MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING} to False."
```

File: mlflow/tracking/fluent.py [hunk 7]
```diff
@@ -380,6 +380,18 @@ def start_run(
             # if `log_system_metrics` is not specified, we will check environment variable.
             log_system_metrics = MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING.get()
         if log_system_metrics:
+            # Ensure psutil is installed. It was moved to an optional dependency, as it doesn't
+            # have binary for Arm64 Linux and requires build from CPython which was a headache.
```
Comment (harupy) [inline mlflow/tracking/fluent.py:384]:
```suggestion
            # have binary for Arm64 Linux and requires build from CPython which is a headache.
```

`was a headache` sounds `it is not now`.

File: mlflow/tracking/fluent.py [hunk 8]
```diff
@@ -380,6 +380,18 @@ def start_run(
...
+            # https://github.com/giampaolo/psutil/issues/1972
+            if importlib.util.find_spec("psutil") is None:
+                raise MlflowException(
+                    "Failed to start system metrics monitoring as package `psutil` is not "
+                    "installed. Run `pip install psutil` to resolve the issue, "
+                    "otherwise you can disable system metrics logging by passing "
+                    "`log_system_metrics=False` to start_run() or setting environment "
```
Comment (harupy) [inline mlflow/tracking/fluent.py:391]:
```suggestion
                    "`log_system_metrics=False` to `start_run()` or setting environment "
```

**Pull request merged**